File: Sorting/MergeSort.cpp
#include<iostream>
#include<vector>

using namespace std;


void Merge(vector<int>&vect,int left,int mid,int right)
{
    int lp=left;
    int rp=mid;
    int pointer=lp;
    vector<int>tvect(right-left+1);
    while((lp<mid) & (rp<=right))
    {
     //   cout<<" lp , mid , rp : "<<lp<<" "<<mid<<" "<<rp<<endl;
        if(vect[lp]<vect[rp])
        {
           tvect[pointer]=vect[lp];
           lp++;
        }
        else if(vect[lp]>vect[rp])
        {
            tvect[pointer]=vect[rp];
            rp++;
        }
        pointer++;
        

    }
    while(lp<mid)
    {
        tvect[pointer]=vect[lp];
        pointer++;
        lp++;
    }
    while(rp<=right)
    {
        tvect[pointer]=vect[rp];
        pointer++;
        rp++;
    }

    for(int i=left;i<=right;i++)
    {
        vect[i]=tvect[i];

    }


}

void MergeSort(vector<int>&vect,int left,int right)
{
    int sze=(right-left+1);
    int midi=left+sze/2;
   // cout<<"size :"<<sze<<endl;  
    if(sze>1)
    {
        
        MergeSort(vect,left,midi-1); //NOTICE how the middle element is considered
        MergeSort(vect,midi,right);

    }
    Merge(vect,left,midi,right);
}



int main()
{
    int n;
    cin>>n;
    vector<int>vect;
    vect.reserve(n);
    int temp;
    for(int i=0;i<n;i++)
    {

        cin>>temp;
        vect.push_back(temp);
    }
  //  cout<<"vect size :"<<vect.size()<<endl;
   
    vector<int>res_vect(n);
     MergeSort(vect,0,n-1);

     cout<<"sorted :"<<endl;

     for(int i=0;i<n;i++)
     {
         cout<<vect[i]<<" ";
     }
     cout<<endl;
    

}

File: Sorting/mergesort.py
'''
TC: O(nlogn)
SC: O(n)

'''
class mergesort():

    def mergesort(self, lst:list, left:int,right:int):
        if(left < right):
            mid = left+(right-left)//2
            self.mergesort(lst, left, mid)
            self.mergesort(lst, mid+1, right) #Handling edge case
            self.merge(lst, left, mid, right)
    
    def merge(self, lst:list, left:int, mid:int, right:int):

        tlst:int = []
        il = left
        ir = mid+1
        while il<=mid and ir<=right:
            if lst[il] <lst[ir]:
                tlst.append(lst[il])
                il+=1
            else:
                tlst.append(lst[ir])
                ir+=1
        
        if il>mid:
            while ir<=right:
                tlst.append(lst[ir])
                ir+=1
        
        if ir>right:
            while il<=mid:
                tlst.append(lst[il])
                il+=1
        
        for it in range(len(tlst)):
            lst[left+it]=tlst[it]



class mergesort2():

    def mergesort(self, lst:list, left:int,right:int):
        if(left < right):
            mid = left+(right-left)//2
            self.mergesort(lst, left, mid)
            self.mergesort(lst, mid+1, right) #Handling edge case
            self.merge(lst, left, mid, right)
    
    def merge(self, lst:list, left:int, mid:int, right:int):

        tlst:int = [0]*(right-left+1)
        cidx = -1
        il = left
        ir = mid+1
        while il<=mid and ir<=right:
            cidx+=1
            if lst[il] <lst[ir]:
                tlst[cidx] = lst[il]
                il+=1
            else:
                tlst[cidx] = lst[ir]
                ir+=1
        
        if il>mid:
            while ir<=right:
                cidx +=1
                tlst[cidx] = (lst[ir])
                ir+=1
        
        if ir>right:
            while il<=mid:
                cidx +=1
                tlst[cidx] = (lst[il])
                il+=1
        
        for it in range(len(tlst)):
            lst[left+it]=tlst[it]

if __name__=="__main__":

    lst = list(map(int,input().split(',')))
    ms = mergesort()
    print(" BEF    :", lst)
    ms.mergesort(lst, 0, len(lst)-1)
    print("Sorted  :", lst)
